src/unix.rs

```diff
@@ -7,27 +7,204 @@
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<(), ()> {
+        // Drain the data from the pipe and maintain interest in getting more
+        let any_wakeup = self.drain();
+        if any_wakeup {
+            self.broadcast();
+        }
+        // This task just lives until the end of the event loop
+        Ok(Async::NotReady)
+    }
+}
 
+impl Drop for Driver {
+    fn drop(&mut self) {
+        let mut drivers = GLOBALS.drivers.lock().unwrap();
+        drivers.remove(&self.id);
+    }
+}
+
+impl Driver {
+    fn new(handle: &Handle) -> Self {
+        Driver {
+            id: handle.id(),
+            // TODO: Any chance of errors here?
+            wakeup: PollEvented::new(EventedReceiver, handle).unwrap(),
+        }
+    }
+    // Drain all data in the pipe and maintain an interest in read-ready
+    fn drain(&self) -> bool {
+        // Inform tokio we're interested in reading. It also hints on
+        // if we may be readable.
+        if let Async::NotReady = self.wakeup.poll_read() {
+            return false;
+        }
+        // Read all available data (until EAGAIN)
+        let mut received = false;
+        let mut buffer = [0; 1024];
+        loop {
+            match recv(GLOBALS.receiver, &mut buffer, MSG_DONTWAIT) {
+                Ok(0) => panic!("EOF on self-pipe"),
+                Ok(_) => received = true,
+                Err(NixError::Sys(Errno::EAGAIN)) => break,
+                Err(NixError::Sys(Errno::EINTR)) => (),
+                Err(e) => panic!("Bad read on self-pipe: {}", e),
+            }
+        }
+
+        // If we got here, it's because we got EAGAIN above. Ask for more data.
+        self.wakeup.need_read();
+
+        received
+    }
+    // Go through all the signals and broadcast everything
+    fn broadcast(&self) {
+        for (sig, value) in GLOBALS.pending.iter().enumerate() {
+            // Any signal of this kind arrived since we checked last?
+            if value.swap(false, Ordering::SeqCst) {
+                let signum = sig as c_int;
+                let mut siginfo = GLOBALS.signals[sig].lock().unwrap();
+                // It doesn't seem to be possible to do this through the iterators for now.
+                // This trick is copied from https://github.com/rust-lang/rfcs/pull/1353.
+                for i in (0 .. siginfo.recipients.len()).rev() {
+                    // TODO: This thing probably generates unnecessary wakups of this task.
+                    // But let's optimise it later on, when we know this works.
+                    match siginfo.recipients[i].start_send(signum) {
+                        // We don't care if it was full or not ‒ we just want to wake up the other
+                        // side.
+                        Ok(AsyncSink::Ready) => {
+                            // We are required to call this if we push something inside
+                            let _ = siginfo.recipients[i].poll_complete();
```

Comment:
These are all channels, so it's ok to avoid `poll_complete` here as we know it doesn't do anything.